tests/proxy/list/list_test.go

```diff
@@ -0,0 +1,388 @@
+}
+
+func TestListProxy_LastIndexOfWithNilElement(t *testing.T) {
+	defer list.Clear()
+	_, err := list.LastIndexOf(nil)
+	AssertErrorNotNil(t, err, "list LastIndexOf() should return error with nil element")
+}
+
+func TestListProxy_Remove(t *testing.T) {
+	defer list.Clear()
+	list.Add("1")
+	removed, err := list.Remove("1")
+	AssertEqualf(t, err, removed, true, "list Remove() failed")
+}
+func TestListProxy_RemoveWithNilElement(t *testing.T) {
+	defer list.Clear()
+	_, err := list.Remove(nil)
+	AssertErrorNotNil(t, err, "list Remove() should return error with nil element")
+}
+
+func TestListProxy_RemoveAt(t *testing.T) {
+	defer list.Clear()
+	list.Add("1")
+	previous, err := list.RemoveAt(0)
+	AssertEqualf(t, err, previous, "1", "list RemoveAt() failed")
+}
+
+func TestListProxy_RemoveAll(t *testing.T) {
+	defer list.Clear()
+	all := []interface{}{"1", "2", "3"}
+	list.AddAll(all)
+	removedAll, err := list.RemoveAll([]interface{}{"2", "3"})
+	AssertEqualf(t, err, removedAll, true, "list RemoveAll() failed")
+	found, err := list.Contains("1")
+	AssertEqualf(t, err, found, true, "list RemoveAll() failed")
+}
+
+func TestListProxy_RemoveAllWithNilElement(t *testing.T) {
+	defer list.Clear()
+	_, err := list.RemoveAll([]interface{}{nil, "1", "2"})
+	AssertErrorNotNil(t, err, "list RemoveAll() should return error with nil element")
+}
+
+func TestListProxy_RemoveAllWithNilSlice(t *testing.T) {
+	defer list.Clear()
+	_, err := list.RemoveAll(nil)
+	AssertErrorNotNil(t, err, "list RemoveAll() should return error with nil slice")
+}
+
+func TestListProxy_RetainAll(t *testing.T) {
+	defer list.Clear()
+	all := []interface{}{"1", "2", "3"}
+	list.AddAll(all)
+	changed, err := list.RetainAll([]interface{}{"2", "3"})
+	AssertEqualf(t, err, changed, true, "list RetainAll() failed")
+	found, err := list.Contains("1")
+	AssertEqualf(t, err, found, false, "list RetainAll() failed")
+}
+
+func TestListProxy_RetainAllWithNilElement(t *testing.T) {
+	defer list.Clear()
+	_, err := list.RetainAll([]interface{}{nil, "1", "2"})
+	AssertErrorNotNil(t, err, "list RetainAll() should return error with nil element")
+}
+
+func TestListProxy_RetainAllWithNilSlice(t *testing.T) {
+	defer list.Clear()
+	_, err := list.RetainAll(nil)
+	AssertErrorNotNil(t, err, "list RetainAll() should return error with nil slice")
+}
+
+func TestListProxy_Size(t *testing.T) {
+	defer list.Clear()
+	all := []interface{}{"1", "2", "3"}
+	list.AddAll(all)
+	size, err := list.Size()
+	AssertEqualf(t, err, size, int32(3), "list Size() failed")
+}
+
+func TestListProxy_Set(t *testing.T) {
+	defer list.Clear()
+	all := []interface{}{"1", "2", "3"}
+	list.AddAll(all)
+	list.Set(1, "13")
+	res, err := list.Get(1)
+	AssertEqualf(t, err, res, "13", "list Set() failed")
+}
+
+func TestListProxy_SetWithNilElement(t *testing.T) {
+	defer list.Clear()
+	_, err := list.Set(0, "")
+	AssertErrorNotNil(t, err, "list Set() should return error with nil element")
+}
+
```

Comment:
We need one more Test:
```
func TestListProxy_SetWithoutItem(t *testing.T) {
	defer list.Clear()
	_, err := list.Set(1, "a")
	AssertErrorNotNil(t, err, "list Set() should return error with index error")
}
```
I tried this test pass but on Invoke method return an err without message. We need something `IndexOutOfBounds`

Reply:
Currently with nonexistent elements server returns an error as the following:
```
		 java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
		 
		 java.util.ArrayList.rangeCheck(ArrayList.java:653)
		 java.util.ArrayList.set(ArrayList.java:444)
		 com.hazelcast.collection.impl.list.ListContainer.set(ListContainer.java:94)
		 com.hazelcast.collection.impl.list.operations.ListSetOperation.run(ListSetOperation.java:61)
		 com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:194)
		 com.hazelcast.spi.impl.operationexecutor.impl.OperationExecutorImpl.run(OperationExecutorImpl.java:406)
		 com.hazelcast.spi.impl.operationexecutor.impl.OperationExecutorImpl.runOrExecute(OperationExecutorImpl.java:433)
		 com.hazelcast.spi.impl.operationservice.impl.Invocation.doInvokeLocal(Invocation.java:569)
		 com.hazelcast.spi.impl.operationservice.impl.Invocation.doInvoke(Invocation.java:554)
		 com.hazelcast.spi.impl.operationservice.impl.Invocation.invoke0(Invocation.java:513)
		 com.hazelcast.spi.impl.operationservice.impl.Invocation.invoke(Invocation.java:207)
		 com.hazelcast.spi.impl.operationservice.impl.InvocationBuilderImpl.invoke(InvocationBuilderImpl.java:60)
		 com.hazelcast.client.impl.protocol.task.AbstractPartitionMessageTask.processMessage(AbstractPartitionMessageTask.java:67)
		 com.hazelcast.client.impl.protocol.task.AbstractMessageTask.initializeAndProcessMessage(AbstractMessageTask.java:123)
		 com.hazelcast.client.impl.protocol.task.AbstractMessageTask.run(AbstractMessageTask.java:103)
		 com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:154)
		 com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:125)
		 com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:100)
		 ------ submitted from ------.(:-1)
		 com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolve(InvocationFuture.java:127)
		 com.hazelcast.spi.impl.AbstractInvocationFuture$1.run(AbstractInvocationFuture.java:243)
		 com.hazelcast.client.impl.protocol.task.AbstractPartitionMessageTask.execute(AbstractPartitionMessageTask.java:78)
		 com.hazelcast.spi.impl.AbstractInvocationFuture.unblock(AbstractInvocationFuture.java:239)
		 com.hazelcast.spi.impl.AbstractInvocationFuture.andThen(AbstractInvocationFuture.java:215)
		 com.hazelcast.client.impl.protocol.task.AbstractPartitionMessageTask.processMessage(AbstractPartitionMessageTask.java:69)
		 com.hazelcast.client.impl.protocol.task.AbstractMessageTask.initializeAndProcessMessage(AbstractMessageTask.java:123)
		 com.hazelcast.client.impl.protocol.task.AbstractMessageTask.run(AbstractMessageTask.java:103)
		 com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:154)
		 com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:125)
		 com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:100)
```
What do you mean by without a message? Do you think this is not sufficient for users ?